### PR TITLE
uart: native_posix: Add support for another serial port

### PR DESCRIPTION
**Pull request by jukkar**
The second serial port (UART_1) can be used for connecting to
host serial port. This is used for example by PPP (Point-to-Point
Protocol) implementation in which case the pppd connects to
the pty that is linked to UART_1.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by jukkar** [comment]
This is just something I needed when implementing ppp support and testing against pppd running in host.

**Comment by aescolar** [comment]
Thanks @jukkar , I will try to take a look ASAP (I have a quite hectic week planned, so it may take a couple of days)

**Comment by jukkar** [comment]
No hurry about this, I just went through all my pending patches and this was one of them. I am not planning to send the PR that needs this anytime soon. But if this is acceptable, then it would be nice to get this one sooner than later.

File: drivers/serial/uart_native_posix.c
Comment (aescolar) [inline drivers/serial/uart_native_posix.c]: Nit: Would a small description be nice here? something like:
```diff
 #if defined(CONFIG_UART_NATIVE_POSIX_PORT_1_ENABLE)
+/*
+ * Initialize the 2nd UART port
+ * This port will be always attached to its own new pseudoterminal
+ */
 static int np_uart_1_init(struct device *dev)
```

Comment (aescolar) [inline drivers/serial/uart_native_posix.c]: Nit: For consistency it feels like the uart 0 init funcion should be renamed:
```diff
diff --git a/drivers/serial/uart_native_posix.c b/drivers/serial/uart_native_posix.c
index 75e53f5f2f..10aeef611a 100644
--- a/drivers/serial/uart_native_posix.c
+++ b/drivers/serial/uart_native_posix.c
@@ -184,13 +188,13 @@ static int open_tty(struct native_uart_status *driver_data,
 }
 
 /**
- * @brief Initialize the native_posix serial port
+ * @brief Initialize the first native_posix serial port
  *
- * @param dev UART device struct
+ * @param dev UART_0 device struct
  *
  * @return 0 (if it fails catastrophically, the execution is terminated)
  */
-static int np_uart_init(struct device *dev)
+static int np_uart_0_init(struct device *dev)
 {
        struct native_uart_status *d;
 
@@ -329,7 +337,7 @@ static int np_uart_tty_poll_in(struct device *dev, unsigned char *p_char)
 }
 
 DEVICE_AND_API_INIT(uart_native_posix0,
-           CONFIG_UART_NATIVE_POSIX_PORT_0_NAME, &np_uart_init,
+           CONFIG_UART_NATIVE_POSIX_PORT_0_NAME, &np_uart_0_init,
            (void *)&native_uart_status_0, NULL,
            PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
            &np_uart_driver_api);
```

File: drivers/serial/uart_native_posix.c
Comment (aescolar) [inline drivers/serial/uart_native_posix.c]: Nit: I guess the explanation in the top would need a minor update (?):
```diff
diff --git a/drivers/serial/uart_native_posix.c b/drivers/serial/uart_native_posix.c
index 75e53f5f2f..10aeef611a 100644
--- a/drivers/serial/uart_native_posix.c
+++ b/drivers/serial/uart_native_posix.c
@@ -20,6 +20,10 @@
 
 /*
  * UART driver for POSIX ARCH based boards.
+ * It can support up to two UARTs.
+ *
+ * For the first UART:
+ *
  * It can either be connected to the process STDIN+STDOUT
  * OR
  * to a dedicated pseudo terminal
```

**Review by aescolar — CHANGES_REQUESTED** [review]
The change seems sensible.
Just a minor thing (kconfig and doc descriptions *1) and a couple of nits
I guess you have tested it works and fulfills your needs?

*1, would it make sense to update the doc to mention this new option?. Something like this(?):
```diff
diff --git a/boards/posix/native_posix/doc/index.rst b/boards/posix/native_posix/doc/index.rst
index 1592e50ff0..6ee606d9f0 100644
--- a/boards/posix/native_posix/doc/index.rst
+++ b/boards/posix/native_posix/doc/index.rst
@@ -536,8 +536,13 @@ The following peripherals are currently provided with this board:
 UART
 *****
 
-This driver can be configured to either create and link the UART to a new
-pseudoterminal (i.e. ``/dev/pts<nbr>``), or to map the UART input and
+This driver can be configured with :option:`CONFIG_UART_NATIVE_POSIX`
+to instantiate up to two UARTs. By default only one UART is enabled.
+With :option:`CONFIG_UART_NATIVE_POSIX_PORT_1_ENABLE`
+you can enable the second one.
+
+For the first UART, it can link it to a new
+pseudoterminal (i.e. ``/dev/pts<nbr>``), or map the UART input and
 output to the executable's ``stdin`` and ``stdout``.
 This is chosen by selecting either
 :option:`CONFIG_NATIVE_UART_0_ON_OWN_PTY` or
@@ -557,8 +562,8 @@ to it. This can be done, for example with the command::
 
 where ``/dev/<ttyn>`` should be replaced with the actual TTY device.
 
-You may also chose to automatically attach a terminal emulator to it by
-passing the command line option ``-attach_uart`` to the executable.
+You may also chose to automatically attach a terminal emulator to the first UART
+by passing the command line option ``-attach_uart`` to the executable.
 The command used for attaching to the new shell can be set with the command line
 option ``-attach_uart_cmd=<"cmd">``. Where the default command is given by
 :option:`CONFIG_NATIVE_UART_AUTOATTACH_DEFAULT_CMD`.
```

**Comment by jukkar** [comment]
> I guess you have tested it works and fulfills your needs?

It works but I want to test this a bit more. I would like to keep this pending a bit longer until I am sure it works as needed. I will send a new version fixing the code as you suggested later.

**Comment by jukkar** [comment]
Updated according to comments. Please do not merge this one yet, I want to do some more testing still.

**Review by aescolar — APPROVED** [review]

**Comment by jukkar** [comment]
Works ok after more testing so removing DNM label. Waiting ack from @dbkinder before merging.

**Comment by jukkar** [comment]
ping @dbkinder, if possible please review

**Comment by jukkar** [comment]
Some issues with data from host to zephyr, investigating.

**Comment by jukkar** [comment]
Rebased against latest master.

**Comment by dbkinder** [comment]
I see @aescolar wanted some documentation updates, but none are included in this PR?

**Comment by jukkar** [comment]
> I see @aescolar wanted some documentation updates, but none are included in this PR?

Good catch, I must have pushed a wrong branch that did not contain the fixes I have made. I will submit a correct branch (but probably tomorrow).

**Comment by jukkar** [comment]
Proper commit submitted, this contains the documentation fixes.

**Pull request merged**
